File: apps/cargo-tracker/src/contexts/billing/domain/repository/invoice-repository.ts
import type { Invoice } from '../model/invoice.js';

/** 請求書リポジトリ（出力ポート） */
export interface InvoiceRepository {
  /** 新規請求書を発行時に永続化する（明細を含む）。二重請求（同一 booking）は UNIQUE 制約で拒否される */
  save(invoice: Invoice): Promise<void>;
  /** 状態遷移・金額・明細を更新する。入金確認済み（CONFIRMED）遷移時は支払記録を追記する */
  update(invoice: Invoice): Promise<void>;
  findByInvoiceId(invoiceId: string): Promise<Invoice | null>;
  findByBookingId(bookingId: string): Promise<Invoice | null>;
  findAll(): Promise<Invoice[]>;
}

File: apps/backend/billingms/src/main/java/com/example/billingms/application/port/InvoiceRepository.java
package com.example.billingms.application.port;

import com.example.billingms.domain.model.Invoice;
import java.util.List;
import java.util.Optional;

/** 精算書の永続化（出力ポート）。 */
public interface InvoiceRepository {

    void save(Invoice invoice);

    Optional<Invoice> findById(String invoiceId);

    List<Invoice> findAll();

    /**
     * その予約に精算書が発行済みか（決定 4）。
     *
     * <p><strong>二重請求を防ぐ。</strong>DB の UNIQUE 制約でも守るが、制約だけだと
     * 画面に 500 が出る——利用者には「なぜ断られたか」が伝わらない。
     */
    boolean existsForBooking(String bookingId);
}
